### PR TITLE
Fixes firelocks preventing doors from closing when activated

### LINKED ISSUE #8300: Fire alarms can be used to open doors you aren't authorized for
This bug came up during the playtest on the 29th of January 2022.
When you turn off a fire alarm, the airlocks open alongside the firebreak doors. This means that if you're quick, you can whack the fire alarm twice and get into an area you're not supposed to.

### PR DESCRIPTION
**Pull request by MaxIsJoe**
Fixes #8300

CL: [Fix] Fixes firelocks preventing doors from closing when activated

### CODE REVIEW

File: UnityProject/Assets/Scripts/Objects/Doors/DoorMasterController.cs [hunk 1]
```diff
@@ -164,6 +164,10 @@ public void UpdateNewPlayer(NetworkConnection playerConn)
 
 		private void TryBump()
 		{
+			if (matrix.GetFirst<FireLock>(registerTile.LocalPositionServer, true) != null)
+			{
+				if(matrix.GetFirst<FireLock>(registerTile.LocalPositionServer, true).fireAlarm.activated) return;
+			}
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/Doors/DoorMasterController.cs]: I don't trust anything in matrix to be cheap. Best to save this to a variable.
```suggestion
			var firelock = matrix.GetFirst<FireLock>(registerTile.LocalPositionServer, true);
			if (firelock != null && firelock.fireAlarm.activated) return;
```

Though it is interesting the state of the fire alarm determines the behaviour here. Wouldn't any closed firedoor need this?

File: UnityProject/Assets/Scripts/Objects/Doors/DoorMasterController.cs [hunk 2]
```diff
@@ -338,6 +341,10 @@ public bool CheckStatusAllow(HashSet<DoorProcessingStates> states)
 
 		public void TryOpen(GameObject originator, bool blockClosing = false)
 		{
+			if (matrix.GetFirst<FireLock>(registerTile.LocalPositionServer, true) != null)
+			{
+				if(matrix.GetFirst<FireLock>(registerTile.LocalPositionServer, true).fireAlarm.activated) return;
+			}
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/Doors/DoorMasterController.cs]:
```suggestion
			var firelock = matrix.GetFirst<FireLock>(registerTile.LocalPositionServer, true);
			if (firelock != null && firelock.fireAlarm.activated) return;
```

File: UnityProject/Assets/Scripts/Objects/Doors/DoorMasterController.cs [hunk 3]
```diff
@@ -492,6 +499,11 @@ public void Close()
 
 		public void Open(bool blockClosing = false)
 		{
+			if (matrix.GetFirst<FireLock>(registerTile.LocalPositionServer, true) != null)
+			{
+				if(matrix.GetFirst<FireLock>(registerTile.LocalPositionServer, true).fireAlarm.activated) return;
+			}
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/Doors/DoorMasterController.cs]:
```suggestion
			var firelock = matrix.GetFirst<FireLock>(registerTile.LocalPositionServer, true);
			if (firelock != null && firelock.fireAlarm.activated) return;
```

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/Matrix.cs [hunk 4]
```diff
@@ -200,8 +201,12 @@ private IEnumerator WaitForNetId()
 	/// </summary>
 	public bool CanCloseDoorAt(Vector3Int position, bool isServer)
 	{
+		if (GetFirst<FireLock>(position, isServer) != null)
+		{
+			if (GetFirst<FireLock>(position, isServer).fireAlarm.activated) return true;
+		}
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/Matrix.cs]:
```suggestion
		var firelock = GetFirst<FireLock>(position, isServer);
		if (firelock != null && firelock.fireAlarm.activated) return true;)
```

**Comment by MaxIsJoe** [comment]
review pls

**Comment by NoooneyDude** [comment]
> Though it is interesting the state of the fire alarm determines the behaviour here. Wouldn't any closed firedoor need this?

Any comment on this?

You can add all suggestions to one commit too, FYI.

**Comment by MaxIsJoe** [comment]
> > Though it is interesting the state of the fire alarm determines the behaviour here. Wouldn't any closed firedoor need this?
> 
> Any comment on this?


I'm not entirely sure what you mean by that, do you mean that firelocks themselves will break because they're checking themselves?
No, because firelocks use the outdated door system and their close/open methods are different from the current V2 doors use.

**Review by NoooneyDude — APPROVED** [review]
Sorry, I misread #8300. Now I'm even more confused... Why is the firedoor bumping the door when closing? With this and #7816, there's something very strange about door code...

Eh, whatever 🙂

**Review by Bod9001 — APPROVED** [review]

**Pull request merged**